boards: st: stm32h7s78_dk: add Ethernet support

bklaric1 opened the pull request:
## Changes

- Added `mac` and `mdio` configuration to `stm32h7s78_dk-common.dtsi` and enabled them. The RMII Ethernet pin configuration follows this schematic from ST: [MB1736](https://www.st.com/resource/en/schematic_pack/mb1736-h7s7l8-d01-schematic.pdf).
- ~Added fallback option for definition of `sram_eth_node` in `mpu_regions.c` and `sections.ld`. This definition earlier defaulted to `sram2`, but now has `sram1` as fallback option.~ ~Defined the SRAM region for Ethernet descriptor and buffer
using `memory-regions = <...>` in the `mac` node (as discussed) and added the updated definition of `sram_eth_node` to `mpu_regions.c` and `sections.ld`. Also added the needed property in `dts/bindings/ethernet/st,stm32h7-ethernet.yaml`.~
  - ~Added `memory-regions = <...>` to `boards/st/nucleo_h7s3l8/nucleo_h7s3l8.dts`, since they use the same `mpu_regions.c` and `sections.ld`.~
- Updated `stm32h7s78_dk.yaml` ~and `stm32h7s78_dk_stm32h7s7xx_ext_flash_app.yaml`~ to include the Ethernet tag.
- Updated `index.rst` to include the Ethernet feature and documented the steps needed in order to use Ethernet for this board.
- Added `tests/drivers/memc/ram/boards/stm32h7s78_dk_stm32h7s7xx_ext_flash_app.overlay` and updated `tests/drivers/memc/ram/boards/stm32h7s78_dk.overlay` since Ethernet drivers fully consume `sram1` node and thus fail the test without the additions.
- ~Added `#memory-region-cells = <0>;` to all possible memory regions (`sram1` and `sram2` nodes) in `dts/arm/st/h7rs/stm32h7rs.dtsi`~

## Tests

I tested these changes with the following examples:

- [dumb_http_server](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/net/sockets/dumb_http_server)
- [http_server](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/net/sockets/http_server)

I could ping the board after flashing both samples and also open the website in the browser. Everything worked as expected.

## Notes

I opened #99164 in order to discuss how to implement the fallback for `sram2` and whether this approach is correct. Taking a look at the discussion should answer some questions about the implementation.

Closes #99164

bklaric1 commented:
The #99505 should be resolved first before proceeding with the merge (if it even gets to that point), since the changes to `mpu_regions.c` might become obsolete.

etienne-lms commented on soc/st/stm32/stm32h7rsx/mpu_regions.c:
```diff
@@ -31,7 +31,13 @@ static const struct arm_mpu_region mpu_regions[] = {
+#define sram_eth_node	DT_NODELABEL(sram2)
+#else
+#define sram_eth_node	DT_NODELABEL(sram1)
+#endif
```
I wonder is an explicit `memory-region = <...>` in the `mac` node would be better to tell which SRAM region is expected to hold ethernet buffers.

bklaric1 replied:
I pushed a commit with this implemented. Thanks for the insight.

bklaric1 commented:
Regarding the failed checks: 

- compliance check fail (intendation and trailing whitespace) will be fixed with the next commit
- twister check error wasn't caused by changes that my commit introduced as far as I can say. The error output:
  ```
  /opt/toolchains/zephyr-sdk-0.17.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/zephyr_pre0.elf section `SRAM1' will not fit in region `SRAM1'
  /opt/toolchains/zephyr-sdk-0.17.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: region `SRAM1' overflowed by 256 bytes
  ```
  states that the `SRAM1` region is overflowed. But the comment in the `mpu_regions.c` file states the following:
  https://github.com/zephyrproject-rtos/zephyr/blob/f1c2ce4f4026bd5b212c6e4cc7d754aba69b189e/soc/st/stm32/stm32h7rsx/mpu_regions.c#L39

So I think this isn't relevant to the PR, but I might be wrong.

Note: next commit will also add `- netif:eth` to [stm32h7s78_dk_stm32h7s7xx_ext_flash_app.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/st/stm32h7s78_dk/stm32h7s78_dk_stm32h7s7xx_ext_flash_app.yaml), since I forgot to add that.

etienne-lms reviewed (commented):
Could you squash to 2 commits?

etienne-lms commented on boards/st/stm32h7s78_dk/stm32h7s78_dk-common.dtsi:
```diff
@@ -211,15 +216,16 @@
 	pinctrl-names = "default";
 	phy-connection-type = "rmii";
 	phy-handle = <&eth_phy>;
+	memory-region = <&sram1>;
```
Your change makes stm32h7rs SoCs to require a `memory-region` property to get the ethernet buffers RAM. nucleo_h7s3l8 board belongs to these SoC series. If not updated, it will fail. It should be updated by a commit before **soc/st/stm32/stm32h7rsx/mpu_regions.c** is updated, or in the very same commit.

etienne-lms commented on dts/bindings/ethernet/st,stm32h7-ethernet.yaml:
```diff
@@ -12,3 +12,10 @@ description: |
+properties:
+  memory-region:
+    type: phandle
+    required: true
```
It is the `required` directive that should be removed.

bklaric1 commented:
Squashed upcoming commit into 2. commit and added the requested changes.

bklaric1 commented:
Regarding the failed [Compliance Check](https://github.com/zephyrproject-rtos/zephyr/actions/runs/19698687522/job/56429386896?pr=99296): I don't see the trailing whitespace that is being reported, neither locally nor in GitHub Web Editor, so I can't really fix this.

bklaric1 commented:
@pdgendt I edited the first commit and removed the trailing whitespace there. Hopefully this fixes the CI check error.

bklaric1 commented:
I think the latter is the better choice here, because we don't want to introduce possible problems with skipping this test for `sram1` node.

I added this change as a third commit. I thought, since it adds changes to unrelated file (to ethernet feature addition) that it should be a separate commit. If wanted, I can squash it into the second commit.

etienne-lms commented:
I forgot to ask a question when starting to review this P-R.:
Why do you want to use `sram1` node for ethernet access instead of `sram2` or done by the nucleo_h7s3l8 board (also stm32h7rs SoC based)? Maybe I missed something.

Digging into the implementation, I wonder if my suggestion to use a `memory-region` DT property is the right way. It may falsely suggest that a board can change this property to locate the ethernet buffers elsewhere. To do so, a bit more changes are needed. Current implementation rely on explicit mapping definition in the SoC specific **mpu_region.c** file.
I've sketched something to see how to handle that in a more flexible way. I'll create an RFC P-R to share.

bklaric1 commented:
I just set it to `sram1` node, since it is enabled by default. I can change it to `sram2` if that is what is prefered.

Yes, both ways this is being done, current implementation and my implementation in `mpu_regions.c` and `sections.ld`, are explicitly mapping ethernet drivers to either `sram1` or `sram2`. 

> I opened #99164 in order to discuss how to implement the fallback for `sram2` and whether this approach is correct. Taking a look at the discussion should answer some questions about the implementation.

As stated here, I myself am unsure if this approach is correct. 

In my opinion, defining what memory region is going to be used for ethernet drivers should be either:

- kept in `mpu_regions.c` and `sections.ld` but then an introduction of a fallback is needed, since by default the `sram2` node is disabled for this board, or explicitly tell the user that in order to use ethernet, the `sram2` node must be enabled (I originally spent quite a while to figure out that it's mapped to `sram2` but `sram2` is disabled by default &rarr; confusing for the user)
- kept like in the current implementation, but tell the user with an inline comment, that `sram1`/`sram2` node is being fully consumed by the ethernet drivers.

> The #99505 should be resolved first before proceeding with the merge (if it even gets to that point), since the changes to `mpu_regions.c` might become obsolete.

In any case, I think this should be addressed first, as using either `sram1` or `sram2` for ethernet drivers might introduce some unwanted problems in the long run:

In any case, I am open to shifting this PR to what your PR agrees on.

bklaric1 commented:
Added the requested change and squashed into one commit (with a long commit message).

etienne-lms commented on soc/st/stm32/stm32h7rsx/mpu_regions.c:
```diff
@@ -31,7 +31,13 @@ static const struct arm_mpu_region mpu_regions[] = {
+
+#if DT_NODE_HAS_COMPAT(sram_eth_node, zephyr_sram)
+#error "Invalid Ethernet memory-region: cannot use zephyr,sram. \
+Use a SRAM region like sram1 or sram2 for Ethernet."
```
```suggestion
BUILD_ASSERT(sram_eth_node != DT_CHOSEN(zephyr_sram)
	     "Remove property memory-region from 'mac' node to locate ethernet buffers in system RAM.")

bklaric1 replied:
@etienne-lms Isn't the assert message exact opposite of what we want to say? The `memory-region` shouldn't be set to `zepyhr,sram` and removing it causes the condition to be false and definition of `sram_eth_node` to be skipped, causing the following error:
```
/opt/toolchains/zephyr-sdk-0.17.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.eth_stm32_desc' from `zephyr/drivers/ethernet/libdrivers__ethernet.a(eth_stm32_hal_common.c.obj)' being placed in section `.eth_stm32_desc'
/opt/toolchains/zephyr-sdk-0.17.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.eth_stm32_buf' from `zephyr/drivers/ethernet/libdrivers__ethernet.a(eth_stm32_hal_common.c.obj)' being placed in section `.eth_stm32_buf'
```

Correct me if I'm wrong

bklaric1 commented:
@etienne-lms Amended the last commit to include the requested changes.

Question: do I need to add a (second) Copyright line when I edit an existing file or only when I create a new file?

bklaric1 commented on dts/bindings/ethernet/st,stm32h7-ethernet.yaml:
```diff
@@ -12,3 +12,9 @@ description: |
 include: st,stm32-ethernet-common.yaml
+
+properties:
+  memory-region:
```
Missed the `s`. Will update.
But if we are trying to be consistent, shouldn't the property also include the `required: true` as in `st,stm32mp13-ethernet.yaml`?
https://github.com/zephyrproject-rtos/zephyr/blob/a1970ebd2f428dc472d505d7c7e5eee4774ee7b0/dts/bindings/ethernet/st%2Cstm32mp13-ethernet.yaml#L16-L22

maass-hamburg commented on soc/st/stm32/stm32h7rsx/mpu_regions.c:
```diff
@@ -30,8 +30,14 @@ static const struct arm_mpu_region mpu_regions[] = {
+
+BUILD_ASSERT(!DT_SAME_NODE(sram_eth_node, DT_CHOSEN(zephyr_sram)),
+	     "Remove property memory-regions from 'mac' node "
+	     "to locate ethernet buffers in system RAM.")
```
you could do
```
#if DT_NODE_HAS_STATUS_OKAY(sram_eth_node) && !DT_SAME_NODE(sram_eth_node, DT_CHOSEN(zephyr_sram))
```
here and in the `sections.ld` file to get rid of the build assert

maass-hamburg commented on dts/bindings/ethernet/st,stm32h7-ethernet.yaml:
```diff
@@ -12,3 +12,9 @@ description: |
 compatible: "st,stm32h7-ethernet"
 
 include: st,stm32-ethernet-common.yaml
```
add `#memory-region-cells = <0>;` to all possible memory regions?

bklaric1 replied:
Adding it fixes the mentioned error and doesn't break anything else with this build (http_server sample). I added it to the `sram0`, `sram1` and `sram2` nodes.
However, I can't guarantee that this addition doesn't break something else unrelated to Ethernet. I can add it to the commit and see if it triggers any twister-build errors. Would that work?

etienne-lms commented:
> Question: do I need to add a (second) Copyright line when I edit an existing file or only when I create a new file

You don't need to add a copyright line at source/header files entry when you modify them to have your copyright implicitly taken into account. What drives the copyright owners list is the Git reversion history, which tracks all contributions, not the file entry inline comment. 
You are free to add your copyright line, but it's not handy when each contributor adds its own line, especially for small modifications.

etienne-lms commented on soc/st/stm32/stm32h7rsx/mpu_regions.c:
```diff
@@ -42,6 +44,12 @@ static const struct arm_mpu_region mpu_regions[] = {
 };
 
+#ifdef sram_eth_node
+	BUILD_ASSERT(!DT_SAME_NODE(sram_eth_node, DT_CHOSEN(zephyr_sram)),
```
Indentation
```suggestion
BUILD_ASSERT(!DT_SAME_NODE(sram_eth_node, DT_CHOSEN(zephyr_sram)),

bklaric1 replied:
Done.

etienne-lms commented on dts/arm/st/h7rs/stm32h7rs.dtsi:
```diff
@@ -54,13 +54,15 @@
 		reg = <0x24000000 DT_SIZE_K(456)>;
 		compatible = "zephyr,memory-region", "mmio-sram";
 		zephyr,memory-region = "SRAM0";
+		#memory-region-cells = <0>;
```
Can be removed. We don't expect one to reference `&sram0` as a `memory-region` phandle.

bklaric1 replied:
Done.

bklaric1 commented on boards/st/nucleo_h7s3l8/nucleo_h7s3l8.dts:
```diff
@@ -181,6 +181,7 @@
 	pinctrl-names = "default";
 	phy-connection-type = "rmii";
 	phy-handle = <&eth_phy>;
+	memory-regions = <&sram1>;
```
`sram2` is disabled by default. That is why I opted to use `sram1`.
I can switch to using `sram2` but that requires the node to be enabled by default. Otherwise an introduction of a fallback is needed (to `sram1`).
What should I do here?
I will update the commit message in any case. Thanks @TD-JBL for clarification.

maass-hamburg replied:
I don't think that preserving behavior is that important for a dev board

erwango reviewed (changes requested):
See [comment for equivalent change on H7](https://github.com/zephyrproject-rtos/zephyr/pull/100910#pullrequestreview-3578364989).
Please apply the same for H7RS (configure memory in .dtsi files)

bklaric1 commented:
@erwango I am going to deal with the requested changes suggested in #100910 first and then apply them to this PR.

bklaric1 committed:
boards: st: stm32h7s78_dk: add Ethernet support

Add Ethernet configuration for mac and mdio nodes in
stm32h78_dk-common.dtsi.

Disable mac node in .overlay files for tests/drivers/memc/ram
since ethernet drivers that get enabled by it consume sram1
node and thus fail the test.

Update the .yaml files to include Ethernet tag and
update documentation to explain Ethernet usage.

Signed-off-by: Benjamin Klaric <benjamin.klaric01@gmail.com>

bklaric1 commented:
@erwango I removed the changes from this PR that have already been added by #101920 and adapted this PR to correspond to what was discussed in #101920.
Do I also need to rebase on main?

erwango reviewed (changes requested):
Thanks!

One comment, otherwise LGTM
